### PR TITLE
Kobo integration

### PR DESCRIPTION
**Pull request by avdata99**
Related to #624 

This PR:
 - Allow users to save their Kobo token as `user plugin_extras`
 - Shows a list of assets after validate token

image

image

image

### CODE REVIEW

File: ckanext/unhcr/utils.py [hunk 1]
```diff
@@ -83,6 +83,20 @@ def view_wrapper(*args, **kwargs):
...
 
+def require_internal_user(func):
+    '''
+    Decorator for flask view functions. Returns 403 response if no user is logged in or if the login user is external
+    '''
+    @wraps(func)
+    def view_wrapper(*args, **kwargs):
+        if (not hasattr(toolkit.c, "user") or not toolkit.c.user):
+            return toolkit.abort(403, "Forbidden")
+        if not is_saml2_user(toolkit.c.userobj):
```
Comment (chris48s) [inline ckanext/unhcr/utils.py:94]: What we're fundamentally trying to work out here is whether the user is "internal" or "external". I possibly confused the issue in #624 by explicitly saying "SAML2".

You'll note a few lines up in this file we have a function `user_is_external()` which works out if a user is internal or external using the `ckanext.unhcr.internal_domains` setting. We implemented this before switching to `keitaroinc/ckanext-saml2auth`.

I think we should have one standardized way of working out if a user is "internal" or "external" and use it everywhere. IMO it is a useful property from a testing perspective (both manual and automated) if we can have "internal" users who are not SAML2 (i.e: it is useful that I can have a variety of different users in my test environment with different permissions https://github.com/okfn/ckanext-unhcr#create-development-users that are not linked to Active Directory but are "internal").  That said, one nice thing is if a user passes `is_saml2_user()` we can rely on `plugin_extras` being a dict that is not `None`, which leads us toward using `if user_is_external(toolkit.c.userobj)` here.

One possible thing we could do to make things easier is just amend `user_create` to always set up `.plugin_extras['unhcr']` and do a data migration to amend legacy users. Then that makes things easier - we can just assume every user has `.plugin_extras['unhcr']`- SAML2 or otherwise. I think that would also apply to users we create under test too.

Maybe this is one for @amercader  to wade in on?

File: ckanext/unhcr/templates/kobo/assets.html [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    <td>{{ asset.asset_type }}</td>
+    <td>{{ asset.owner__username }}</td>
+    <td>
+        {% for download in asset.downloads %}
+        <a href="{{ download.url }}" role="button" class="btn btn-sucess">{{ download.format }}</a>
+        {% endfor %}
+    </td>
+    <td>
+    <div class="btn-group pull-right">
+        <form method="POST" action="{# Import endpoint #}">
```
Comment (chris48s) [inline ckanext/unhcr/templates/kobo/assets.html:32]: TODO

Assume the intent is to do this in a different PR?

File: ckanext/unhcr/tests/pt/test_kobo.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    def test_internal_user_kobo_asset(self, kobo_asset, app):
+        environ = {
+            'REMOTE_USER': self.internal_user['name']
+        }
+
+        asset = self.get_asset(1)
+        assert 'url' in asset
+        assert 'TEST aaZSQ29VRHJmofjGSfDDv4' == asset['name']
+        kobo_asset.return_value = asset
+        # TODO test asset page if exists
```
Comment (chris48s) [inline ckanext/unhcr/tests/pt/test_kobo.py:87]: is this to do in this PR?

File: ckanext/unhcr/actions.py [hunk 4]
```diff
@@ -1205,12 +1205,16 @@ def user_show(up_func, context, data_dict):
...
+    unhcr_extras = _validate_plugin_extras(plugin_extras['unhcr'])
+    user['focal_point'] = unhcr_extras['focal_point']
+    user['expiry_date'] = unhcr_extras['expiry_date']
+    user['default_containers'] = unhcr_extras['default_containers']
 
-    user['focal_point'] = extras['focal_point']
-    user['expiry_date'] = extras['expiry_date']
-    user['default_containers'] = extras['default_containers']
+    # CKAN only allow showing plugin extras for sysadmins
+    # TODO validate we want to allow this
```
Comment (chris48s) [inline ckanext/unhcr/actions.py:1215]: think this one is a question for @amercader  too

File: ckanext/unhcr/templates/kobo/assets.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        <input type="hidden" value="{{ asset.uid }}" name="kobo_asset_id" />
+        <button
+            type="submit"
+            class="btn btn-primary"
+            title="Import Kobo Asset to CKAN"
+        >
+        Import to CKAN
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/assets.html]: Fixed [here](https://github.com/okfn/ckanext-unhcr/commit/70296430d9f77d78bb303fa92a9f64dea0add9e7)

File: ckanext/unhcr/templates/kobo/assets.html [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+{% block primary_content %}
+
+<h1>Kobo Assets</h1>
+
+<table class="table table-header table-hover table-bordered">
+<thead>
+    <tr>
+    <th>Name</th>
+    <th>Type</th>
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/assets.html]: Refactor assets -> surveys [here](https://github.com/okfn/ckanext-unhcr/commit/bd9e0f293cbe384b42fa6d2052dd55629d64f029)

File: ckanext/unhcr/templates/kobo/assets.html [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    <th>User</th>
+    <th>Download</th>
+    <th>Import</th>
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/assets.html]: Proposed to change [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/2393a181c685eb638977608fd8b89a167f4e59aa)

File: ckanext/unhcr/templates/kobo/index.html [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ <a target='_blank' href="https://kobo.unhcr.org/">log in to UNHCR Kobo instance</a>
+and then 
+<a target='_blank' href="https://kobo.unhcr.org/token/?format=json">get your Kobo token</a></p>
+<p>You will see something like this:</p>
+<pre>
+  {"token":"abc123def456"}</pre>
+<p>Your token will be used to authenticate to Kobo and get you data from Kobo.</p>
+
+<form action="{% url_for 'unhcr_kobo.kobo_assets' %}" method="POST">
+<input type="text" name="kobo_token" value="{{ kobo_token }}" size="45" maxlength="120"/>
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/index.html]: Fixed [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/581fac100b8a8a9bab3447f3926ed5ad86539d87)

File: ckanext/unhcr/templates/kobo/base.html [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+{% block secondary_content %}
+  <h2 class="module-heading"><i class="fa fa-line-chart"></i> Kobo integration</h2>
+  
+  <nav>
+    <ul class="unstyled nav nav-simple">
+      <li class="nav-item">
+        <a target='_blank' href="https://{{ h.kobo_url() }}/token/?format=json">Get my Kobo token</a>
+      </li>
+      {% if h.kobo_token %}
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/base.html]: Fixed [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/72217f2db1e604f11b25b3b1dcf0046a6bd70517)

File: ckanext/unhcr/tests/factories.py [hunk 10]
```diff
@@ -51,3 +51,19 @@ class ExternalUser(factories.User):
...
+    plugin_extras = {
+        'saml2auth': {
+            'saml_id': "saml_id_{}".format(user.name)
+        }
+    }
+
+    return plugin_extras
+
+
+class AzureUser(factories.User):
```
Comment (amercader) [inline ckanext/unhcr/tests/factories.py:67]: See comment below about user types. I think this should be called `InternalUser`

File: ckanext/unhcr/utils.py (hunk 1 above)
Comment (amercader) [reply ckanext/unhcr/utils.py:94]: Creating an empty `.plugin_extras['unhcr']` on all users sounds good as well

**Review by amercader — COMMENTED** [review]
Great work @avdata99, I added some comments to follow up

File: ckanext/unhcr/actions.py [hunk 11]
```diff
@@ -1260,11 +1264,29 @@ def user_update(up_func, context, data_dict):
...
                 up_dict['apikey'] = data_dict.get('apikey')
-                return up_func(context, up_dict)
+
+            old_kobo_token = up_dict['plugin_extras'].get('unhcr', {}).get('kobo_token')
+            new_kobo_token = data_dict.get('plugin_extras', {}).get('unhcr', {}).get('kobo_token')
+            kobo_token_changed = new_kobo_token and old_kobo_token != new_kobo_token
+            if kobo_token_changed:
+                if 'unhcr' not in up_dict['plugin_extras']:
+                    up_dict['plugin_extras']['unhcr'] = {}
+                up_dict['plugin_extras']['unhcr']['kobo_token'] = new_kobo_token
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/actions.py:1291]: Changed [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/c22c5dc698cc7f99e6a85c96d765eba60e9703a3). I will also redirect to `/bad-token` url

File: ckanext/unhcr/templates/kobo/base.html [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+      <li class="nav-item">
+        <a href="{{ h.url_for('unhcr_kobo.index') }}" title="Change my Kobo token">Change my Kobo token</a>
+      </li>
+      <li class="nav-item">
+        <a href="{{ h.url_for('unhcr_kobo.kobo_assets') }}" title="Change my Kobo token">My Kobo assets</a>
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/base.html]: Fixed

File: ckanext/unhcr/templates/kobo/index.html [hunk 13]
```diff
@@ -0,0 +1,23 @@
+{% extends "kobo/base.html" %}
+
+{%- block subtitle %}Kobo{% endblock -%}
+
+{% block primary_content %}
+
+<h1>Kobo Integration</h1>
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/index.html]: Changed [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/9a077cfe6db0eded88651626854137ca4694dc31)

File: ckanext/unhcr/templates/kobo/assets.html [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+</thead>
+<tbody>
+{% for asset in assets %}
+<tr>
+    <td><b>{{ asset.name }}</b></td>
+    <td>{{ asset.asset_type }}</td>
+    <td>{{ asset.owner__username }}</td>
+    <td>
+        {% for download in asset.downloads %}
+        <a href="{{ download.url }}" role="button" class="btn btn-sucess">{{ download.format }}</a>
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/assets.html]: Links dropped [here](https://github.com/okfn/ckanext-unhcr/commit/e37f7d6bccc8ecf6f2934a578f59cf3baba48f9b)

File: ckanext/unhcr/kobo/api.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        self.token = token
+        self.base_url = 'https://{}/api/v2/'.format(base_domain)
+        self.assets = None
+
+    def _get(self, resource_url):
+        """ Get any api/v2 resource in JSON format """
+        url = resource_url if resource_url.startswith('http') else self.base_url + resource_url
+        try:
+            response = requests.get(url, headers={'Authorization': 'Token ' + self.token})
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/kobo/api.py]: Updated [here](https://github.com/okfn/ckanext-unhcr/commit/a9f0e3f8b40540c49ab5884d764b9105bfcb2c2e)

File: ckanext/unhcr/kobo/api.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+            raise e
+
+        return response.json()
+
+    def get_assets(self, limit=100, force=False):
+        # TODO allow pagination
+        if self.assets is None or force:
+            response = self._get(resource_url='assets.json?limit={}'.format(limit))
+            self.assets = response['results']
+            while 'next' in response and response['next'] is not None:
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/kobo/api.py]: Simplified [here](https://github.com/okfn/ckanext-unhcr/commit/00056f97414c1d45323d9c03b53b4b109525d746)

File: ckanext/unhcr/utils.py (hunk 1 above)
Comment (avdata99) [reply ckanext/unhcr/utils.py]: I agree that we need to improve how user types are defined. It is not very clear now.
I think the best way to handle it is through a new ticket. I can create it using these comments. 
I created the [AzureUser factory](https://github.com/okfn/ckanext-unhcr/blob/624_kobo_integration/ckanext/unhcr/tests/factories.py#L66-L67) (maybe it should be called Saml2User) to simplify some tests.
Abount this specific change: do you think [changing just this](https://github.com/okfn/ckanext-unhcr/commit/f385da819ef5825689f9985e0317c6de6b313a9e) it's ok for this PR?

File: ckanext/unhcr/templates/kobo/assets.html (hunk 2 above)
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/assets.html]: Yes @chris48s, this should be covered in #625 
I [hid the button](https://github.com/okfn/ckanext-unhcr/commit/79e94ad8adfcb194507afc8b3c64e1c2cffed9f9) until next PR

File: ckanext/unhcr/tests/pt/test_kobo.py (hunk 3 above)
Comment (avdata99) [reply ckanext/unhcr/tests/pt/test_kobo.py]: [Removed](https://github.com/okfn/ckanext-unhcr/commit/79a9618b67e1b52487c49264484ad9b5eaf92b9b), will be applied next PR

File: ckanext/unhcr/actions.py (hunk 4 above)
Comment (avdata99) [reply ckanext/unhcr/actions.py]: If the user is a sysadmin `include_plugin_extras` should work fine. I [propose to remove it](https://github.com/okfn/ckanext-unhcr/commit/1b34ea189aaced182d25380871f8ed2f38fe4748). I was trying to get the Kobo token without requiring a sysadmin.
I will validate this and update tests to ensure it work as expected.
Assuming only sysadmin can read and update `plugin_extras` maybe I will need to review more on this PR

File: ckanext/unhcr/blueprints/kobo.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    __name__,
+    url_prefix=u'/kobo'
+)
+
+
+@require_internal_user
+def index():
+    user_obj = toolkit.c.userobj
+    kobo_token = user_obj.plugin_extras.get('unhcr', {}).get('kobo_token')
+    context = {
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/blueprints/kobo.py]: Fixed [here](https://github.com/okfn/ckanext-unhcr/commit/ee80b5a0442855018761a2504ac19256ffa9bf27)

File: ckanext/unhcr/blueprints/kobo.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        'kobo_token': kobo_token
+    }
+    return toolkit.render('kobo/index.html', context)
+
+
+@require_internal_user
+def kobo_bad_token():
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/blueprints/kobo.py:28]: By mistake, I stopped using this endpoint in a [previous commit](https://github.com/okfn/ckanext-unhcr/pull/635#discussion_r674163223). 
It's back [now](https://github.com/okfn/ckanext-unhcr/commit/de3a9105cd8141ab53970965611e3f605e6426ae).

File: ckanext/unhcr/actions.py [hunk 19]
```diff
@@ -1260,11 +1265,36 @@ def user_update(up_func, context, data_dict):
         user_obj = m.User.get(user_id)
 
         if user_obj is not None and is_saml2_user(user_obj):
-
-            if data_dict.get('apikey') and user_obj.apikey != data_dict.get('apikey'):
-                up_dict = toolkit.get_action('user_show')(context.copy(), {'id': user_id})
+            # Only allow to change apiKey or kobo_token
+            new_context = context.copy()
+            new_context['include_plugin_extras'] = True
+            up_dict = toolkit.get_action('user_show')(new_context, {'id': user_id})
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/actions.py:1266]: Old kobo token is not loaded using the model `plugin_extras` 
Since we don't use a sysadmin here (should we?) I load all plugin extras from model [here](https://github.com/okfn/ckanext-unhcr/commit/a3c60352035ccf2c80863ef248c6fa6409b5f910).
Is this a good idea?
Should I use `package_revise` for the update?

[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/actions.py:1266]: @amercader I check how the `saml2auth` extension update the `plugin_extras` [here](https://github.com/keitaroinc/ckanext-saml2auth/blob/main/ckanext/saml2auth/views/saml2auth.py#L158-L169) and it seem you don't care about `plugin_extras` from other extensions.
I'm not sure which is the best way to update `plugin_extras`

File: ckanext/unhcr/actions.py [hunk 20]
```diff
@@ -1260,11 +1265,36 @@ def user_update(up_func, context, data_dict):
...
+                        "Kobo token is not valid"
+                    ]})
+                if 'unhcr' not in up_dict['plugin_extras']:
+                    up_dict['plugin_extras']['unhcr'] = {}
+                up_dict['plugin_extras']['unhcr']['kobo_token'] = new_kobo_token
+
+            if apikey_changed or kobo_token_changed:
+                new_context = context.copy()
+                # Required to update plugin_extras
+                new_context['ignore_auth'] = True
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/actions.py]: I mistakenly assumed that it was necessary to correct the authorization in order to update `plugin_extras`.
Fixed [here](https://github.com/okfn/ckanext-unhcr/commit/6d78039b1077d14eea2a89e142a5f6dd67a99465).

File: ckanext/unhcr/blueprints/kobo.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/blueprints/kobo.py:28]: It's now used [here](https://github.com/okfn/ckanext-unhcr/blob/624_kobo_integration/ckanext/unhcr/blueprints/kobo.py#L53)

File: ckanext/unhcr/helpers.py [hunk 21]
```diff
@@ -241,16 +241,35 @@ def user_is_curator(userobj=None):
     return user_id in user_ids
 
 
-def user_is_container_admin(user=None):
-    if not user:
-        user = toolkit.c.user
+def user_orgs(permission, userobj=None):
+    if not userobj:
+        userobj = toolkit.c.user
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/helpers.py]: Updated [here](https://github.com/okfn/ckanext-unhcr/commit/e6199c9b13db86d9c808da32e0cb6f9c191fb5c0)

File: ckanext/unhcr/helpers.py [hunk 22]
```diff
@@ -241,16 +241,35 @@ def user_is_curator(userobj=None):
...
+    orgs = user_orgs(permission='create_dataset', userobj=userobj)
+    return len(orgs) > 0
+
+
+def get_kobo_token():
+    user = toolkit.c.userobj
+    return user.plugin_extras.get('unhcr', {}).get('kobo_token')
+
+
+def get_kobo_url():
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/helpers.py]: Updated [here](https://github.com/okfn/ckanext-unhcr/commit/0a26663929784aa65768d72dc8cd1ec5439232f0)

File: README.md [hunk 23]
```diff
@@ -223,3 +223,9 @@ We use a fake SMTP server to test email notifications:
 - restart the development server
 
 Now all email sent by `from ckan.lib.mailer import mail_user` should be sent to the `Demo Inbox` at Mailtrap.
+
+## Kobo integration
+
+Internal users are allowed to import data from [KoboToolbox](https://www.kobotoolbox.org/#home).  
+To define where you Kobo instance lives you can configure the `ckanext.unhcr.kobo_domain` setting (default is `kobo.unhcr.org`).
```
[the remark this answers is not in the document]
Comment (avdata99) [reply README.md]: Updated [here](https://github.com/okfn/ckanext-unhcr/commit/2f993541b03d9db3a079b0e255fa7d054fd132ce)
(same as last comment)

File: ckanext/unhcr/kobo/api.py [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+
+    def get_surveys(self, limit=100, force=False, all_pages=True):
+        if self.assets is None or force:
+            response = self._get(resource_url='assets.json?limit={}'.format(limit))
+            self.assets = [asset for asset in response['results'] if asset['asset_type'] == 'survey']
+            if all_pages:
+                while 'next' in response and response['next'] is not None:
+                    response = self._get(resource_url=response['next'])
+                    new_results = [asset for asset in response['results'] if asset['asset_type'] == 'survey']
+                    self.assets += new_results
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/kobo/api.py]: Chris already mention this code. It was changed to simplify it.

File: ckanext/unhcr/templates/kobo/index.html [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+<p>As a RIDL user you could authenticate to Kobo using your own credentials.</p> 
+<p>This means you will be able to import surveys and data from Kobo.</p>
+<p>To achieve that you will need to 
+ <a target='_blank' href="https://kobo.unhcr.org/">log in to UNHCR Kobo instance</a>
+and then 
+<a target='_blank' href="https://kobo.unhcr.org/token/?format=json">get your Kobo token</a></p>
+<p>You will see something like this:</p>
+<pre>
+  {"token":"abc123def456"}</pre>
+<p>Your token will be used to authenticate to Kobo and get you data from Kobo.</p>
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/index.html]: Fixed [here](https://github.com/okfn/ckanext-unhcr/commit/11c19b812a01b627d25b547c8b680845162a66f1) (you have a typo)

File: ckanext/unhcr/templates/kobo/surveys.html [hunk 26]
```diff
@@ -0,0 +1,71 @@
+{% extends "kobo/base.html" %}
+
+{%- block subtitle %}Kobo{% endblock -%}
+
+{% block primary_content %}
+
+<h1>Kobo Surveys</h1>
+
+{% if not h.user_is_editor %}
+You can't import any of these surveys because you are not editor in any data container.
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/surveys.html]: I changed to only allow those endpoint only to editors [here](https://github.com/okfn/ckanext-unhcr/commit/a0150488956f3ff7173fde7a19dc28bb36cf998e)

File: ckanext/unhcr/templates/kobo/index.html (hunk 13 above)
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/index.html]: and [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/240f2ce1a9411c41b08886e5396de3a1a8e4ad34)

File: ckanext/unhcr/tests/factories.py (hunk 10 above)
Comment (avdata99) [reply ckanext/unhcr/tests/factories.py]: Renamed [here](https://github.com/okfn/ckanext-unhcr/commit/30d6c49779eac4d84c88f83bbfdcc44a074bc893)

File: ckanext/unhcr/utils.py (hunk 1 above)
Comment (chris48s) [reply ckanext/unhcr/utils.py]: > If we actually need to check a user comes from AD to deal with specific SAML2 stuff then we can use `is_saml2_user()` (in fact I suspect that our current usage of `is_saml2_user()` should use `not user_is_external()`)

Prior to this PR the situation where we are using `is_saml2_user()` is to prevent SAML2 users from editing their profile details or requesting a password reset. So in that case I think we do actually want to check specifically whether the user is managed by SAML2. A user who is "internal" but "native" should still be able to do those things.

> @chris48s @avdata99 What do you think?

Beyond that, yes I agree that we should mostly use `user_is_external()` / `user_is_internal()` based on `ckanext.unhcr.internal_domains`as far as possible just because for local dev it is more convenient to have a bunch of test users with different permissions, memberships, etc. If your local dev account has to be a SAML account to test certain features then you only have one account to play with.

File: ckanext/unhcr/templates/kobo/bad_token.html [hunk 27]
```diff
@@ -0,0 +1,12 @@
+{% extends "kobo/base.html" %}
```
Comment (amercader) [inline ckanext/unhcr/templates/kobo/bad_token.html:1]: I really don't like this page, I find it really confusing. IIUC there are two potential errors:
1. The token submitted is not correct (maybe users pasted an old one, paster the whole `{"token": "xxx"}` string, missed a letter, etc) In this case the error should be shown in the same page where the form was submitted 
2. There was an error while performing a request to the KoBo API. The only API request we are doing now is getting the user surveys. If there is an error here, it should be displayed in the surveys page

I know that we are doing the same request to check 1 and 2, but users don't know that and when they are doing 1 (setting their token) it is really confusing to be redirected to a page that says "There was an error, maybe it's your token maybe it's the API, you decide what to do" :slightly_smiling_face: 

To show the errors you can add a `<div class="error-explanation alert alert-error"></div>` like the dataset form  or use `h.flash_error()`:
Screenshot 2021-07-26 at 12-55-10 Create Dataset - RIDL

In fact, if we are doing 2 (getting the surveys) after the token was successfully set (ie `token=None`) there is [no exception handling from the API](https://github.com/okfn/ckanext-unhcr/pull/635/files#diff-19702f802f993ff75e3c918373bd7ebf1b8e1c0c511a82e815ce38be83309bdfR67) at all right?

I think the best will be to split this `surveys()` endpoint into two separate ones, with separate exception handling:
1. POST to create/update the token (and clear, more on this later!). This can be a `POST` request to `unhcr_kobo.index`. If successful, redirects to `unhcr_kobo.surveys`
2. GET request to get the surveys (`unhcr_kobo.surveys`)

We can then get rid of the bad_token page.

File: ckanext/unhcr/templates/kobo/index.html [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+          name="kobo_token"
+          value=""
+          placeholder="Insert your KoBo token here"
+          size="45"
+          maxlength="120"
+          class="form-control"/>
+    </div>
+  </div>
+  <div class="form-actions">
+    <input type="submit" class="btn btn-primary" value="{% if kobo_token %}Update{% else %}Submit{% endif %} token" />
```
Comment (amercader) [inline ckanext/unhcr/templates/kobo/index.html:39]: We need an option to clear the token (ie deleting it). I tried to submit an empty field and it didn't work, and in any case a "Remove token" button here will be more intuitive.

**Review by amercader — COMMENTED** [review]
We are getting there @avdata99, just some more work needed in the frontend and we can call this done

File: ckanext/unhcr/templates/kobo/surveys.html [hunk 29]
```diff
@@ -0,0 +1,63 @@
+{% extends "kobo/base.html" %}
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/surveys.html:1]: We can handle this surveys table here. 
#637 should handle imports.

File: ckanext/unhcr/templates/user/dashboard.html [hunk 30]
```diff
@@ -16,6 +16,17 @@
       {% set icon = 'exclamation' if total else 'check' %}
       {{ h.build_nav_icon('unhcr_dashboard.requests', _('Pending Requests') + suffix, icon=icon) }}
     {% endif %}
+
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/unhcr/templates/user/dashboard.html:19]: Added [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/2312afeaea308cb8868689861f1a13b8ba0e4b6a)
image

File: ckanext/unhcr/templates/kobo/bad_token.html (hunk 27 above)
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/bad_token.html]: Improve token error [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/081e8ab510bf2a02141ccbbc6dfdc0b3db3336e5)
Selection_282

Comment (avdata99) [reply ckanext/unhcr/templates/kobo/bad_token.html]: An `update-token` and `remove-token` endpoints [were added](https://github.com/okfn/ckanext-unhcr/pull/635/commits/7029e07e0377764c880c851b16d12eab6aa7e171)
I think it is clearer now.

File: ckanext/unhcr/templates/kobo/index.html (hunk 28 above)
Comment (avdata99) [reply ckanext/unhcr/templates/kobo/index.html:39]: Now we have a new _Remove Token_ button
image

Tests are green but I'm having some troubles locally. Still debugging them.

File: ckanext/unhcr/blueprints/kobo.py [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+            else:
+                message = 'Unknown KoBo token error: {}.'.format(e)
+            toolkit.h.flash_error(message)
+            return toolkit.redirect_to('unhcr_kobo.index')
+
+    return toolkit.redirect_to(redirect_on_success)
+
+
+def remove_token(redirect_on_success='unhcr_kobo.index'):
+    """ POST a new token """
```
Comment (amercader) [inline ckanext/unhcr/blueprints/kobo.py:56]:
```suggestion
    """ Delete existing KoBo token """
```

File: ckanext/unhcr/actions.py [hunk 32]
```diff
@@ -1273,14 +1273,19 @@ def user_update(up_func, context, data_dict):
...
-                # check if the token is valid
-                kobo_url = toolkit.config.get('ckanext.unhcr.kobo_url', 'https://kobo.unhcr.org')
-                kobo = KoBoAPI(new_kobo_token, kobo_url)
-                if not kobo.test_token():
-                    raise toolkit.ValidationError({'kobo_token': [
-                        "KoBo token is not valid"
-                    ]})
-                up_dict['plugin_extras'] = user_obj.plugin_extras
+                if new_kobo_token == 'REMOVE':
+                    new_kobo_token = ''
```
Comment (amercader) [inline ckanext/unhcr/actions.py]: Can we use `None` here?

File: ckanext/unhcr/blueprints/kobo.py [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+        toolkit.get_action('user_update')(
+            {'user': user_obj.name},
+            {'id': user_obj.id, 'plugin_extras': {'unhcr':  {'kobo_token': 'REMOVE'}}}
+        )
+    except toolkit.ValidationError as e:
+        message = 'Error removing token: {}.'.format(e)
+        toolkit.h.flash_error(message)
+        return toolkit.redirect_to('unhcr_kobo.index')
+
+    return toolkit.redirect_to(redirect_on_success)
```
Comment (amercader) [inline ckanext/unhcr/blueprints/kobo.py:69]: Can we not use the route name directly (`unhcr_kobo.index`) here instead of using a variable? AFAICT we are not passing a different value from elsewhere

File: ckanext/unhcr/blueprints/kobo.py [hunk 34]
```diff
@@ -0,0 +1,123 @@
...
+            )
+        except toolkit.ValidationError as e:
+            if e.error_dict and 'kobo_token' in e.error_dict:
+                message = 'Invalid KoBo token. Fix it and try again.'
+            else:
+                message = 'Unknown KoBo token error: {}.'.format(e)
+            toolkit.h.flash_error(message)
+            return toolkit.redirect_to('unhcr_kobo.index')
+
+    return toolkit.redirect_to(redirect_on_success)
```
Comment (amercader) [inline ckanext/unhcr/blueprints/kobo.py:52]: Can we not use the route name directly (`unhcr_kobo.kobo_surveys`) here instead of using a variable? AFAICT we are not passing a different value from elsewhere

File: ckanext/unhcr/blueprints/kobo.py [hunk 35]
```diff
@@ -0,0 +1,123 @@
...
+
+    if token != old_token:
+        try:
+            toolkit.get_action('user_update')(
+                {'user': user_obj.name, 'validate_token': not token.startswith('test_')},
+                {'id': user_obj.id, 'plugin_extras': {'unhcr':  {'kobo_token': token}}}
+            )
+        except toolkit.ValidationError as e:
+            if e.error_dict and 'kobo_token' in e.error_dict:
+                message = 'Invalid KoBo token. Fix it and try again.'
```
Comment (amercader) [inline ckanext/unhcr/blueprints/kobo.py:46]:
```suggestion
                message = 'Invalid KoBo token. Please try again.'
```

File: ckanext/unhcr/blueprints/kobo.py [hunk 36]
```diff
@@ -0,0 +1,123 @@
...
+    plugin_extras = {} if user_obj.plugin_extras is None else user_obj.plugin_extras
+    token = plugin_extras.get('unhcr', {}).get('kobo_token')
+
+    # check if the token is valid
+    kobo_url = toolkit.config.get('ckanext.unhcr.kobo_url', 'https://kobo.unhcr.org')
+    kobo = KoBoAPI(token, kobo_url)
+
+    try:
+        surveys = kobo.get_surveys()
+    except Exception as e:
```
Comment (amercader) [inline ckanext/unhcr/blueprints/kobo.py:86]: We really need to avoid catch-all Exceptions like this one. I suggest creating a `KoboException` class in the `kobo/api.py` module that we raise explicitly when needed (ie just [here](https://github.com/okfn/ckanext-unhcr/pull/635/files#diff-7177da887f4e624e91e757a37fef4883258d73484fe9a58a4fb0639d19debe5eR25) right now) and catching it here instead of `Exception`. Otherwise we can mask other unrelated failures.

**Review by amercader — COMMENTED** [review]
Nice work @avdata99 . Some minor changes and we are good to go

File: ckanext/unhcr/actions.py (hunk 32 above)
Comment (avdata99) [reply ckanext/unhcr/actions.py]: This is outdated, we now [remove this key](https://github.com/okfn/ckanext-unhcr/blob/624_kobo_integration/ckanext/unhcr/actions.py#L1283)

File: ckanext/unhcr/blueprints/kobo.py (hunk 33 above)
Comment (avdata99) [reply ckanext/unhcr/blueprints/kobo.py]: Changed [here](https://github.com/okfn/ckanext-unhcr/commit/f1209e90c6566b1a8ea781b71ba26d63e254d149)

File: ckanext/unhcr/blueprints/kobo.py (hunk 34 above)
Comment (avdata99) [reply ckanext/unhcr/blueprints/kobo.py]: Updated [here](https://github.com/okfn/ckanext-unhcr/commit/d3f3136896c7dca7bcc6589afa6fb0aa54f7a7b6)

File: ckanext/unhcr/blueprints/kobo.py (hunk 36 above)
Comment (avdata99) [reply ckanext/unhcr/blueprints/kobo.py]: Added Kobo error [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/002a3b5bfedcb3af4fbdbd0ce0498eed2e2f29b9).

**Review by amercader — APPROVED** [review]

**Comment by amercader** [comment]
Great work @avdata99 this part is good to go
